### PR TITLE
Added the baseline model training code, changed the README of baselin…

### PR DESCRIPTION
**Pull request by SkylerZheng**
…e model accordingly.

*Issue #, if available:*

*Description of changes:*
Added the baseline model training code, changed the README of baseline model accordingly. 

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/VAL_product_search/README.md [hunk 1]
```diff
@@ -39,28 +39,50 @@ git clone --recurse-submodules https://github.com/alexa/amazon-alexa-language-as
...
 	```
 	cp -r ./VAL/* ./VAL_product_search/
 	```
 
-#### Evaluation offline
+#### Training
+
+- Go to VAL\_product\_search directory, use the following bash command for model training:
+	```
+	bash scripts/run_product_search.sh
```
Comment (spongezhang) [inline src/VAL_product_search/README.md]: change the name to run_product_search_train.sh?

File: src/VAL_product_search/README.md [hunk 2]
```diff
@@ -39,28 +39,50 @@ git clone --recurse-submodules https://github.com/alexa/amazon-alexa-language-as
...
-	```
-	python download_tool/data_download.py -o ./VAL_product_search/datasets/data/product_search -i ./VAL_product_search/datasets/data/query_file_released.jsonl -t qe_jsonl
-	```
+	 - To download the query images for evaluation, download the evaluation gallery csv file (query\_file\_released.jsonl to ./src/VAL\_product\_search/datasets/data) and run following code in ./src
+		```
+		python download_tool/data_download.py -o ./VAL_product_search/datasets/data/product_search -i ./VAL_product_search/datasets/data/query_file_released.jsonl -t qe_jsonl
+		```
 
-* Download the trained VAL checkpoints from [Amazon Drive](https://www.amazon.com/clouddrive/share/TssaYesoP9ux2NpdEkorB0P7ikI7HyB1iLYcG3S8S6l), and put them under VAL\_product\_search/save\_model/product\_search/val\_resnet\_v2\_50\_ml/checkpoints/ directory.
+* 3. Download the trained VAL checkpoints (for evaluation) from [Amazon Drive](https://www.amazon.com/clouddrive/share/TssaYesoP9ux2NpdEkorB0P7ikI7HyB1iLYcG3S8S6l), and put them under VAL\_product\_search/save\_model/product\_search/val\_resnet\_v2\_50\_ml/checkpoints/ directory.
```
Comment (spongezhang) [inline src/VAL_product_search/README.md]: Are you sure we don't need to change the name of the checkpoints here?

File: src/VAL_product_search/datasets/product_search.py [hunk 3]
```diff
@@ -68,23 +68,28 @@ def generate_queries_(self):
...
+      if not os.path.isfile(source_filename):
+        print("Missing source file: ", source_filename)
         continue
-      self.source_files += [filename]
-      self.target_files += [os.path.join(self.path, line[1])]
+      self.source_files += [source_filename]
+      self.target_files += [target_filename]
       line[2] = line[2].strip()
       self.modify_texts += [self.caption_post_process(s=line[2])]
+      self.num_modifiable_imgs += 1
```
Comment (spongezhang) [inline src/VAL_product_search/datasets/product_search.py]: Not clear what does num_modifiable_imgs mean

File: src/VAL_product_search/train.py [hunk 4]
```diff
@@ -0,0 +1,295 @@
...
+FLAGS = tf.app.flags.FLAGS
+
+# Functions used for model building
+squeeze = lambda x: tf.squeeze(x, [1, 2])
+normalization = lambda x: tf.nn.l2_normalize(x, 1, 1e-10)
+expand_dims = lambda x: tf.expand_dims(tf.expand_dims(x, 1), 1)
+pairwise_distance = lambda f1, f2, dim: tf.reduce_sum(tf.square(tf.subtract(f1, f2)), dim)
+
+def _build_model(source_images, target_images, modify_texts, seqlengths, vocab_size):
+    """ define model & loss """
```
Comment (spongezhang) [inline src/VAL_product_search/train.py]: Add docstring for arguments.

File: src/VAL_product_search/datasets/product_search.py (hunk 3 above)
Comment (SkylerZheng) [reply src/VAL_product_search/datasets/product_search.py]: This counts the number of the overall training samples, "modifiable" comes from the original code, we can change it to self.num_queries. This variable will be used in training stage.

File: src/VAL_product_search/README.md (hunk 1 above)
Comment (SkylerZheng) [reply src/VAL_product_search/README.md]: run_product_search_train.sh makes more sense, will do.

File: src/VAL_product_search/train.py (hunk 4 above)
Comment (SkylerZheng) [reply src/VAL_product_search/train.py]: sure

File: src/VAL_product_search/README.md (hunk 2 above)
Comment (spongezhang) [inline src/VAL_product_search/README.md]: nit: pretrained

File: src/VAL_product_search/README.md [hunk 5]
```diff
@@ -39,28 +39,50 @@ git clone --recurse-submodules https://github.com/alexa/amazon-alexa-language-as
...
 
-* Download the data annotation files from [Amazon Drive](https://www.amazon.com/clouddrive/share/KceftgFr8Ii07ONZHHCAc7qGNTyRsS20rP3oLE2ASkN), and put them under VAL\_product\_search/datasets/product\_search/captions\_pairs directory.
+* 4. Download the data annotation files from [Amazon Drive](https://www.amazon.com/clouddrive/share/KceftgFr8Ii07ONZHHCAc7qGNTyRsS20rP3oLE2ASkN), and put them under VAL\_product\_search/datasets/product\_search/captions\_pairs directory.
+	
+* 5. Download the pre-trained glove word embeddings to initialize the text model from [Amazon Drive](https://www.amazon.com/clouddrive/share/hbZgPxjw5tmlYjgzztSeIjSnApWM2LWJ4MsqatWcxPe), and put them under VAL\_product\_search/glove directory.
 
-* Copy all the files under VAL submodule to src/VAL\_product\_search directory.
+* 6. Download the ImageNet pretrained model [resnet](http://download.tensorflow.org/models/resnet_v2_50_2017_04_14.tar.gz), and put it under VAL\_product\_search/pretrained\_model/resnet\_v2\_50 directory.
```
Comment (spongezhang) [inline src/VAL_product_search/README.md:68]: nit: create the dir and make a placeholder file for the commit?

File: src/VAL_product_search/README.md (hunk 2 above)
Comment (SkylerZheng) [reply src/VAL_product_search/README.md:53]: I think we can keep the current names. But if to change, what names do you suggest?

Comment (SkylerZheng) [reply src/VAL_product_search/README.md]: Changed as suggested, thanks!

File: src/VAL_product_search/extract_features.py [hunk 6]
```diff
@@ -15,9 +15,9 @@
 from extract_features_val import _text_model, _image_model, _image_modify_model
 
 tf.app.flags.DEFINE_string(
-  'dev_filename', 'datasets/product_search/captions_pairs/product_search-cap-dev_gallery_images.txt', 'path to the annotation files for feature extractions for gallery images and queries')
+  'dev_filename', 'datasets/product_search/captions_pairs/product_search_gallery_images.txt', 'path to the annotation file of gallery images for feature extractions')
```
Comment (spongezhang) [inline src/VAL_product_search/extract_features.py:18]: We also need to update the file name here: https://www.amazon.com/clouddrive/share/KceftgFr8Ii07ONZHHCAc7qGNTyRsS20rP3oLE2ASkN/folder/E4anmQX-SvaU72xhsdN-Ag ?

File: src/VAL_product_search/extract_features.py [hunk 7]
```diff
@@ -15,9 +15,9 @@
...
 tf.app.flags.DEFINE_string(
-  'train_filename', 'datasets/product_search/captions_pairs/product_search-cap-train_3_feedbacks.txt', 'path to the annotation file for vocabulary extraction')
+  'train_filename', 'datasets/product_search/captions_pairs/product_search_train.txt', 'path to the training annotation file for vocabulary extraction')
```
Comment (spongezhang) [inline src/VAL_product_search/extract_features.py:20]: We also need to update the file name here: https://www.amazon.com/clouddrive/share/KceftgFr8Ii07ONZHHCAc7qGNTyRsS20rP3oLE2ASkN/folder/E4anmQX-SvaU72xhsdN-Ag ?

**Review by spongezhang — APPROVED** [review]
